### PR TITLE
check annotations and paths

### PR DESCRIPTION
**Pull request by JimBugwadia**
Signed-off-by: Jim Bugwadia <jim@nirmata.com>

Fixes https://github.com/kyverno/policies/issues/301

## Description

Sample policies to restrict NGINX Ingress annotation values and Ingress host paths

## Checklist

- [x] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [x] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.

### CODE REVIEW

**Comment by JimBugwadia** [comment]
@chipzoller - did you find a good source for the host path values to restrict? 

The rules file

```	invalidEtcDir         = regexp.MustCompile(`/etc/(passwd|shadow|group|nginx|ingress-controller)`)
	invalidSecretsDir     = regexp.MustCompile(`/var/run/secrets`)

File: best-practices/nginx_path_checks/restrict-path.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    policies.kyverno.io/severity: high
+    policies.kyverno.io/subject: Ingress, NGINX Ingress
+    policies.kyverno.io/minversion: 1.6.0
+    policies.kyverno.io/description: |-
+      This policy mitigates CVE-2021-25745 by restricting `spec.rules[].http.paths[].path` to safe values.
+spec:
+  rules:
+    - name: check-paths
+      match:
+        resources:
```
Comment (chipzoller) [inline best-practices/nginx_path_checks/restrict-path.yaml:17]: Let's put this under either an `any` or `all` block to be consistent with conventions.

File: best-practices/nginx_annotation_checks/restrict-annotations.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    policies.kyverno.io/subject: Ingress, NGINX Ingress
+    policies.kyverno.io/minversion: 1.6.0
+    policies.kyverno.io/description: |-
+      This policy mitigates CVE-2021-25746 by restricting use of metadata.annotations.values to safe values
+      as specified in https://github.com/kubernetes/ingress-nginx/blame/main/internal/ingress/inspector/rules.go.
+spec:
+  rules:
+    - name: check-annotation-values
+      match:
+        resources:
```
Comment (chipzoller) [inline best-practices/nginx_annotation_checks/restrict-annotations.yaml:18]: Let's put this under either an `any` or `all` block to be consistent with conventions.

**Review by chipzoller — CHANGES_REQUESTED** [review]
Comments, plus can we add the `kyverno-version` and `kubernetes-version` annotations as documented [here](https://github.com/kyverno/policies/wiki/Kyverno-annotations)? When I wrote and tested, I did so on Kyverno 1.6.2 and Kubernetes 1.23.

**Comment by chipzoller** [comment]
> @chipzoller - did you find a good source for the host path values to restrict?
> 
> The rules file has 2 checks:
> 
> https://github.com/kubernetes/ingress-nginx/blame/main/internal/ingress/inspector/rules.go
> 
> ```go
>         invalidEtcDir         = regexp.MustCompile(`/etc/(passwd|shadow|group|nginx|ingress-controller)`)
>         invalidSecretsDir     = regexp.MustCompile(`/var/run/secrets`)
> ```
> 
> So wondering if we should provide additional or limit to that.

In the path rule I just filtered for `/etc/` to make it simple. And `/var/run/secrets` is covered. Or did you mean something else?

**Comment by JimBugwadia** [comment]
There are a number of other path checks at:

https://github.com/kyverno/policies/pull/302/files#diff-057a9bd58ec1015a0ead70c75ed14a6eb413c1df84a912d82ee3ad86b91ca701R25

Can we remove them, only keeping `/etc` and `var/run/secrets`?

**Comment by chipzoller** [comment]
Sorry, I can't see the diff on mobile.

**Comment by JimBugwadia** [comment]
Looks like some of the prior policy tests are failing:

applying 1 policy to 3 resources... 

│───│─────────────────────────────│──────────────────────────────────│─────────────────────────────────────│────────│
│ # │ POLICY                      │ RULE                             │ RESOURCE                            │ RESULT │
│───│─────────────────────────────│──────────────────────────────────│─────────────────────────────────────│────────│
│ 1 │ cert-manager-limit-duration │ certificate-duration-max-100days │ default/Certificate/letsencrypt-crt │ Fail   │
│ 2 │ cert-manager-limit-duration │ certificate-duration-max-100days │ default/Certificate/acme-crt-short  │ Pass   │
│ 3 │ cert-manager-limit-duration │ certificate-duration-max-100days │ default/Certificate/acme-crt-long   │ Pass   │
│───│─────────────────────────────│──────────────────────────────────│───────────────────────────────

There may be others...

**Comment by chipzoller** [comment]
Yes, there's a CLI code regression causing this. I pointed this out a little while ago. It's not a policy problem.

**Comment by realshuting** [comment]
> Yes, there's a CLI code regression causing this. I pointed this out a little while ago. It's not a policy problem.

@mritunjaysharma394 - was this related to your recent change of re-structuring CLI?

@vyankyGH  - can you help fix the CLI failure?

**Comment by chipzoller** [comment]
> There are a number of other path checks at:
> 
> https://github.com/kyverno/policies/pull/302/files#diff-057a9bd58ec1015a0ead70c75ed14a6eb413c1df84a912d82ee3ad86b91ca701R25
> 
> Can we remove them, only keeping `/etc` and `var/run/secrets`?

I see the diff now. I was going by mitigation suggestion [here](https://github.com/kubernetes/ingress-nginx/issues/8502) where it says check against newly added rules or that ConfigMap list. I just combined them together to be extra safe.

**Comment by JimBugwadia** [comment]
> I see the diff now. I was going by mitigation suggestion https://github.com/kubernetes/ingress-nginx/issues/8502 where it says check against newly added rules or that ConfigMap list. I just combined them together to be extra safe.

I believe that applies to annotations as well as paths. I reduced the path list to remove the annotation directives.

We can submit and get more feedback from the Kubernetes authors.

**Comment by chipzoller** [comment]
From my reading of the [issue](https://github.com/kubernetes/ingress-nginx/issues/8502), which applies to both CVEs (so the path and the annotation), the same checks need to apply to both. It says rules or ConfigMap, so I just did both to be safe. 

Rules:
```
	invalidAliasDirective = regexp.MustCompile(`\s*alias\s*.*;`)
	invalidRootDirective  = regexp.MustCompile(`\s*root\s*.*;`)
	invalidEtcDir         = regexp.MustCompile(`/etc/(passwd|shadow|group|nginx|ingress-controller)`)
	invalidSecretsDir     = regexp.MustCompile(`/var/run/secrets`)
	invalidByLuaDirective = regexp.MustCompile(`.*_by_lua.*`)
```

ConfigMap:

```
"load_module,lua_package,_by_lua,location,root,proxy_pass,serviceaccount,{,},',\""
```

Maybe for the annotations policy we should add a note advising that the [ConfigMap](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/#annotation-value-word-blocklist) is also recommended separately.

Also, would you mind adding those Kyverno and Kubernetes annotations showing on which versions this was successfully tested?

**Review by chipzoller — APPROVED** [review]

**Pull request merged**